broke-broker/src/main/java/org/dsngroup/broke/broker/ServerContext.java

```diff
@@ -46,6 +53,71 @@ public MessagePool getMessagePool() {
+        for (ServerSession serverSession: serverSessionPool.asCollection()) {
+            // Only publish to active sessions
+            if(serverSession.getIsActive()) {
+                // Whether the mqttPublish Message matches the subscription is performed in "publishToSubscription"
+                serverSession.publishToSubscription(mqttPublishMessage);
+            }
+        }
+    }
+
+    /**
+     * TODO: group based publish
+     * 1. Get all the sessions with matched subscription(s)
+     * 2. Group these sessions with group ID
+     * 3. For each group, perform selection algorithm
+     * 4. For each group, publish the message to the selected subscriber(session)
+     * */
+    public void groupBasedPublishToSubscription(MqttPublishMessage mqttPublishMessage) {
+        String topic = mqttPublishMessage.variableHeader().topicName();
+        TreeMap<Integer, ArrayList<ServerSession>> groupIdSessionListMap = new TreeMap<>();
+        for(ServerSession serverSession: serverSessionPool.asCollection()) {
+            if(serverSession.getIsActive()) {
+                SubscriptionPool subscriptionPool = serverSession.getSubscriptionPool();
+                Subscription matchedSubscription = subscriptionPool.getMatchSubscription(topic);
+                if(matchedSubscription!=null) {
+                    int groupId = matchedSubscription.getGroupId();
+                    if(!groupIdSessionListMap.containsKey(groupId)) {
+                        groupIdSessionListMap.put(groupId, new ArrayList<>());
+                    }
+                    groupIdSessionListMap.get(groupId).add(serverSession);
+                }
+            }
+        }
+        for(Map.Entry<Integer, ArrayList<ServerSession>> entry: groupIdSessionListMap.entrySet()) {
+            int groupId = entry.getKey();
+            ArrayList sessionList = entry.getValue();
+            // Select over the session lists
+            ServerSession selectedSession = selectSession(sessionList);
+            logger.debug("Group ID: "+groupId+" Publish to "+selectedSession.getClientId());
+            selectedSession.publishToSubscription(mqttPublishMessage);
+        }
+    }
+
+    private ServerSession selectSession(ArrayList<ServerSession> sessionList) {
+        double[] scoreArray = new double[sessionList.size()];
+        for(int i=0; i<sessionList.size();i++) {
+            scoreArray[i] = sessionList.get(i).getPublishScore();
+        }
+        double maxScore = 0;
+        int maxScoreIdx = 0;
+        for(int i=0; i<scoreArray.length; i++) {
+            if(maxScore<scoreArray[i]){
+                maxScore = scoreArray[i];
+                maxScoreIdx = i;
+            }
+            // TODO: delete this debug message.
+            // logger.debug("Client ID: "+sessionList.get(i).getClientId()+ " Score: "+scoreArray[i]+" \n");
+        }
+        return sessionList.get(maxScoreIdx);
+    }
+
```

Comment:
I think it's time to split out the scoring and session related methods into separated classes, instead of `ServerContext`